### ISSUE TITLE
[Ray Java] can't transfer ObjectRef from JAVA application to PYTHON deployment

### ISSUE THREAD

**Issue by yucai** [ray-project/ray#45676]
### What happened + What you expected to happen

Reproduction steps:

1. Compile and run the provided JavaCallPython class.
```
java -classpath /mount/data/yucai/ray-demo-1.0-SNAPSHOT-jar-with-dependencies.jar -Dray.address=127.0.0.1:6379 -Dray.job.code-search-path=/mount/data/yucai/ JavaCallPython
```
2. Observe "Can't transfer JAVA data to PYTHON" error.

Expected Behavior:
Pass ObjectRef from Java to Python should be supported.

Actual Behavior:
We met "Can't transfer JAVA data to PYTHON" error When an ObjectRef is passed as a parameter. 

### Versions / Dependencies

Ray: 2.9.0

### Reproduction script

```
import io.ray.api.ObjectRef;
import io.ray.api.Ray;
import io.ray.serve.api.Serve;
import io.ray.serve.deployment.Application;
import io.ray.serve.generated.DeploymentLanguage;
import io.ray.serve.handle.DeploymentHandle;

/*
app@nsfw-ray-yucai-head-lf78j:/mount/data/yucai$ cat test_python_deployment.py
from ray import serve

@serve.deployment
class Counter(object):
    def __init__(self, value):
        self.value = int(value)

    def do_test(self, data):
        return str(self.value)
 */
public class JavaCallPython {

    public static void main(String[] args) throws Exception {
        Serve.start(null);

        Application deployment = Serve.deployment()
                .setLanguage(DeploymentLanguage.PYTHON)
                .setName("JavaCallPython")
                .setDeploymentDef("test_python_deployment.Counter")
                .setNumReplicas(1)
                .bind(28);
        DeploymentHandle handle = Serve.run(deployment).get();

        for (int i = 0; i < 1000; i++) {
            StringBuilder sb = new StringBuilder(20000000);
            for (int n = 0; n < 20000000; n++) {
                sb.append('a');
            }
            String data = sb.toString();
            ObjectRef<String> dataRef = Ray.put(data);

            System.out.println(handle.method("do_test").remote(dataRef).result());
            data = null;

            System.out.println("iter: " + i);
            Thread.sleep(500);
        }
    }
}
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by yucai** [comment]
```
app@nsfw-ray-dev-head-kzqrk:/opt/tritonserver$ java -classpath /mount/data/yucai/ray-demo-1.0-SNAPSHOT-jar-with-dependencies.jar -Dray.address=127.0.0.1:6379 -Dray.job.code-search-path=/mount/data/yucai/ JavaCallPython
2024-06-03 04:24:12,699 INFO Serve [main]: Got controller handle with name `SERVE_CONTROLLER_ACTOR` in namespace `serve`.
2024-06-03 04:24:12,746 INFO Serve [main]: Connecting to existing Serve app in namespace serve
2024-06-03 04:24:12,773 INFO Serve [main]: Connecting to existing Serve app in namespace serve
2024-06-03 04:24:12,795 INFO MetricRegistry [main]: Finished startup metric registry, metricConfig is MetricConfig{timeIntervalMs=5000, threadPoolSize=1, shutdownWaitTimeMs=3000}.
2024-06-03 04:24:17,914 INFO ServeControllerClient [main]: Deployment 'JavaCallPython' is ready . component=serve deployment=JavaCallPython
2024-06-03 04:24:18,252 INFO LongPollClientFactory [main]: Serve.getReplicaContext()` may only be called from within a Ray Serve deployment.
2024-06-03 04:24:18,255 INFO LongPollClientFactory [main]: LongPollClient was initialized
2024-06-03 04:24:18,255 INFO LongPollClientFactory [main]: LongPollClient registered keys: [(RUNNING_REPLICAS,JavaCallPython)].
2024-06-03 04:24:18,255 INFO LongPollClientFactory [ray-serve-long-poll-client-thread]: LongPollClient polls next snapshotIds {(RUNNING_REPLICAS,JavaCallPython)=-1}
2024-06-03 04:24:18,280 INFO LongPollClientFactory [ray-serve-long-poll-client-thread]: LongPollClient received updates for keys: [(RUNNING_REPLICAS,JavaCallPython)]
2024-06-03 04:24:18,286 INFO ReplicaSet [ray-serve-long-poll-client-thread]: ReplicaSet: +1, -0 replicas.
2024-06-03 04:24:18,286 INFO LongPollClientFactory [main]: LongPollClient polls next snapshotIds {(RUNNING_REPLICAS,JavaCallPython)=798420}
2024-06-03 04:24:19,287 INFO LongPollClientFactory [main]: Register poll timeout. keys:[(RUNNING_REPLICAS,JavaCallPython)]
Exception in thread "main" java.lang.IllegalArgumentException: Can't transfer JAVA data to PYTHON
        at io.ray.runtime.task.ArgumentsBuilder.wrap(ArgumentsBuilder.java:72)
        at io.ray.runtime.AbstractRayRuntime.callActorFunction(AbstractRayRuntime.java:369)
        at io.ray.runtime.AbstractRayRuntime.callActor(AbstractRayRuntime.java:208)
        at io.ray.api.call.PyActorTaskCaller.remote(PyActorTaskCaller.java:32)
        at io.ray.serve.router.ReplicaSet.tryAssignReplica(ReplicaSet.java:116)
        at io.ray.serve.router.ReplicaSet.assignReplica(ReplicaSet.java:80)
        at io.ray.serve.router.Router.assignRequest(Router.java:83)
        at io.ray.serve.handle.DeploymentHandle.remote(DeploymentHandle.java:91)
        at JavaCallPython.main(JavaCallPython.java:41)
```

**Comment by yucai** [comment]
When an ObjectRef is passed as a parameter, the Java serializer (FstSerializer) is employed for data serialization, thereby setting cross-language compatibility to 'False'. Is this an intentional design choice to restrict the use of ObjectRef?

```
      String message = "Hello, Ray!";
      ObjectRef<String> objectRef = new ObjectRefImpl<>();
      Object[] foo = new Object[] {"hello", objectRef};
      Pair<byte[], Boolean> serialized = Serializer.encode(foo);
```
<img width="1771" alt="image" src="https://github.com/ray-project/ray/assets/2989575/1cc09006-e00e-4584-87d2-da00a956b6f2">

**Comment by yucai** [comment]
@jjyao I created a PR to support ObejctRef transfer when cross language and I have tested in our production already.
Can someone help preview and guide us to fix it?

**Comment by yucai** [comment]
My PR: https://github.com/ray-project/ray/pull/45729

**Comment by yucai** [comment]
@GeneDer , could you help review my PR?

**Comment by yucai** [comment]
This issue is currently a significant roadblock preventing us from moving our key Ray-based use case into a production environment. @jjyao could you priority this issue? We have PR already, if anyone can guide us, we can fix together.

**Comment by GeneDer** [comment]
Thanks for the contribution @yucai. The code LGTM, but I'm not be best person to review it and nor a code owner. Will wait for other's review

**Comment by yucai** [comment]
Thank you so much, @GeneDer! I am awaiting feedback from the expert at AntGroup. :)

**Comment by jjyao** [comment]
Closed by #45729
